SimPoints warmup

studyztp opened the pull request:
Modify the SimPoints warmup section. It now doesn't check the gap between SimPoints, instead, it checks if the length of the SimPoint start_insts is shorter than the warmup interval. If it is, then it takes the SimPoint start_insts as the warmup interval length. 
Added a simpoint_save_checkpoint_generator to take checkpoints for SimPoints. It will output checkpoint dir name as "cpt.SimPoint[index]".

studyztp committed:
stdlib: modify SimPoints warmup and add a SimPoints generator

In simpoint.py, delected variable last, which was used to store the
last SimPoint start_inst so we know if the gap between two SimPoints
is smaller than the warmup interval. Right now, the warmup only checks
if the start_insts length is smaller than the warmup interval.

In exit_event_generators.py, added a
simpoints_save_checkpoint_generator() to specificlly take checkpoints
for SimPoints. The checkpoint names will be "cpt.SimPoint{index}".
It will also take multiple checkpoints for SimPoints that stop at the
same instruction counts.

studyztp committed:
base, stdlib: ensure the same inst count only schedule once

In base.cc, added a variable last to store the last inst count and make
sure that the same inst counts don't get to schedule more than once.

In simpoint.py, added a detection to make sure that all start_insts are
non-zero.

powerjg commented on src/cpu/base.cc:
```diff
@@ -727,8 +727,12 @@ void
 {
     std::string cause = "simpoint starting point found";
+    Counter last = -1;
     for (size_t i = 0; i < inst_starts.size(); ++i) {
-        scheduleInstStop(0, inst_starts[i], cause);
+        if (last != inst_starts[i]) {
+            scheduleInstStop(0, inst_starts[i], cause);
+            last = inst_starts[i];
+        }
```
I suggest a different approach:
- Check for duplicates with `const bool hasDuplicates = std::adjacent_find(inst_starts.begin(), inst_starts.end()) != inst_starts.end();
`
- Raise a fatal if there are duplicates
- Remove duplicates in python before calling this function.

studyztp replied:
The duplicates should be allowed because the warmup length can be larger than multiple SimPoint starting instructions. When it happens, the ```inst_starts``` for them will all be 0 and cause the duplicates.

powerjg replied:
I don't think that's what the code is doing... if you allow duplicates then why are you checking `last != inst_starts[i]`?

studyztp replied:
The code prevents scheduling exit events for duplicate instructions count, but it allows the list to have duplicates.

powerjg replied:
Why do you think it's better to not remove the duplicates in python? 

I have to say, I don't like increasing the complexity of this code. I think that it would be better to not allow duplicates, but I'm open to having my mind changed.

studyztp replied:
For example, when the SimPoints = [1, 2, 3] with SimPoint interval = 1000 and warmup length = 2000, then the `inst_starts` = [0, 0, 1000]. I want the code only schedule one exit event for instruction 0 and let the  `simpoints_save_checkpoint_generator` to deal with saving checkpoints for SimPoints with the same `inst_start`.

studyztp replied:
I can remove the duplicates in the [base_cpu_core.py](https://gem5.googlesource.com/public/gem5/+/refs/heads/develop/src/python/gem5/components/processors/base_cpu_core.py#155) instead of here, so it doesn't increase the complexity in the C++ code and allow the list in the stdlib SimPoint module to have duplicate starting instructions.

powerjg commented on src/python/gem5/simulate/exit_event_generators.py:
```diff
@@ -143,3 +144,24 @@ def save_checkpoint_generator(checkpoint_dir: Path):
+        m5.checkpoint((checkpoint_dir / f"cpt.SimPoint{count}").as_posix())
+        last_start = simpoint_list[count]
+        count += 1
+        while(count < len(simpoint_list) and 
```
Can you add a comment to explain why this while loop is needed?

studyztp replied:
Added comments :)

powerjg commented on src/python/gem5/utils/simpoint.py:
```diff
@@ -94,6 +94,11 @@ def __init__(
+        if (0 in self._simpoint_start_insts):
+            for index in range(len(self._simpoint_start_insts)):
+                if(self._simpoint_start_insts[index]==0):
+                    self._simpoint_start_insts[index]=1
```
Why 1?

studyztp replied:
I worried about scheduling exit events for instruction 0, but scheduling events for instruction 0 is the same as scheduling them for instruction 1. I deleted this part.

powerjg commented on src/python/gem5/utils/simpoint.py:
```diff
@@ -94,6 +94,11 @@ def __init__(
+        if (0 in self._simpoint_start_insts):
+            for index in range(len(self._simpoint_start_insts)):
+                if(self._simpoint_start_insts[index]==0):
+                    self._simpoint_start_insts[index]=1
```
Why 1?

powerjg replied:
Why 1?

powerjg reviewed (commented):
Mostly small questions below that could use some comments.

studyztp committed:
stdlib: added comments for generator

also removed the block that changes start_insts to 1 when it's 0 because
it's not needed.

studyztp commented:
@powerjg @BobbyRBruce Sorry for the late reply. Thank you for reviewing them! The fixed files are committed :) 
Edit: Sorry, forgot to @

powerjg commented:
Yeah, that is what I was suggesting :)

On Fri, Oct 7, 2022 at 12:48 PM studyztp ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/cpu/base.cc
> <https://github.com/darchr/gem5/pull/251#discussion_r990454498>:
>
> > +    Counter last = -1;
>      for (size_t i = 0; i < inst_starts.size(); ++i) {
> -        scheduleInstStop(0, inst_starts[i], cause);
> +        if (last != inst_starts[i]) {
> +            scheduleInstStop(0, inst_starts[i], cause);
> +            last = inst_starts[i];
> +        }
>
> I can remove the duplicates in the base_cpu_core.py
> <https://gem5.googlesource.com/public/gem5/+/refs/heads/develop/src/python/gem5/components/processors/base_cpu_core.py#155>
> instead of here, so it doesn't increase the complexity in the C++ code and
> allow the list in the stdlib SimPoint module to have duplicate starting
> instructions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/darchr/gem5/pull/251#discussion_r990454498>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA2YHAPP7GN6ZN3AWJDMV3WCB5CHANCNFSM6AAAAAAQ37YK2U>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Jason Lowe-Power (he/him/his)
Assistant Professor, Computer Science Department
University of California, Davis
3049 Kemper Hall
https://arch.cs.ucdavis.edu/

studyztp commented:
This is better than my original approach. You are way ahead of me lol. I'll fix the codes :)

studyztp committed:
stdlib, cpu: remove duplicates in SimPoints list

In base_cpu_core.py, remove the duplicates in the inst_starts before passing
it into scheduling functions for SimPoint exit events.
Added some extra comments for set_simpoint().

studyztp committed:
stdlib: removed the temp variable

studyztp commented:
@powerjg @BobbyRBruce I removed the changes in the baseCPU and added codes to remove duplicates in the `inst_starts` before passing the list into the scheduling functions in baseCPU.

powerjg commented on src/python/gem5/components/processors/base_cpu_core.py:
```diff
@@ -155,9 +155,9 @@ def get_mmu(self) -> BaseMMU:
     def set_simpoint(self, inst_starts: List[int], init: bool) -> None:
         if init:
-            self.core.simpoint_start_insts = inst_starts
+            self.core.simpoint_start_insts = [*set(inst_starts)]
```
I'm not sure there's any guarantee that this is in order. It does seem to be in practice, but I can't find where in the documentation it is guaranteed. 

```suggestion
            self.core.simpoint_start_insts = sorted(set(inst_starts))
```

studyztp replied:
Yes, you're right. I changed it to using sorted().

studyztp committed:
stdlib: use sorted() instead of [] to convert set to list
